geo/src/algorithm/triangulate_spade.rs

```diff
@@ -0,0 +1,710 @@
+use geo_types::{Coord, Line, Point, Triangle};
+use spade::{
+    ConstrainedDelaunayTriangulation, DelaunayTriangulation, Point2, SpadeNum, Triangulation,
+};
+
+use crate::{
+    line_intersection::line_intersection, CoordsIter, EuclideanDistance, GeoFloat,
+    LineIntersection, LinesIter,
+};
+use crate::{Centroid, Contains};
+
+// ====== Error ========
+
+#[derive(Debug)]
+pub enum TriangulationError {
+    SpadeError(spade::InsertionError),
+    LoopTrap,
+    ConstraintFailure,
+}
+
+impl std::fmt::Display for TriangulationError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl std::error::Error for TriangulationError {}
+
+pub type TriangulationResult<T> = Result<T, TriangulationError>;
+
+// ======= Float trait ========
+
+pub trait SpadeTriangulationFloat: GeoFloat + SpadeNum {}
+impl<T: GeoFloat + SpadeNum> SpadeTriangulationFloat for T {}
+
+// ======= Triangulation trait =========
+
+pub type Triangles<T> = Vec<Triangle<T>>;
+
+// seal the trait that needs to be implemented for TriangulateSpade to be implemented. This is done
+// so that we don't leak these weird methods on the public interface.
+mod private {
+    use super::*;
+    pub trait TriangulationRequirementTrait<'a, T>
+    where
+        T: SpadeTriangulationFloat,
+    {
+        /// collect all the lines that are relevant for triangulations from the geometric object that
+        /// should be triangulated.
+        ///
+        /// intersecting lines are allowed
+        fn get_lines(&'a self) -> Vec<Line<T>>;
+        /// collect all the points that are relevant for triangulations from the geometric object that
+        /// should be triangulated
+        fn get_points(&'a self) -> Vec<Coord<T>>;
```

Comment:
Minor perf thing: Can we pass through the underlying iter to avoid allocating a Vec? It seems like it's only usage is here:

```
       let points = self.get_points();
        points
            .into_iter()
```

So it seems like there's no need to collect the iter into a vec since we immediately turn it back into an iter.


It'd be nice if the same applied to `get_lines`, but it seems harder to accomplish because `prepare_intersection_contraint` inserts and removes lines. It still might be possible with filter_map/flatten but it might not be worth the hassle in that case.

Reply:
But I like the idea in general! Maybe I'll have the nerves for this another day 😅

Reply:
I made a PR for this here: https://github.com/RobWalt/geo/pull/1

Because different iterators are involved, it'd need to be a boxed trait (unless/until "return `impl Trait` from trait" ever gets stabilized).